shared/domains/chelonia/chelonia.js

```diff
@@ -829,30 +831,199 @@ export default (sbp('sbp/selectors/register', {
+                  return
+                } else if (state === 'read-new-response' || buffer) {
+                  // If done prematurely, throw an error
+                  controller.error(new Error('Invalid response: done too early'))
+                  return
+                } else {
+                  // If there are still events to fetch, switch state to fetch
+                  state = 'fetch'
+                  break
+                }
+              }
+              if (!value) {
+                // If there's no value (e.g., empty response), throw an error
+                controller.error(new Error('Invalid response: missing body'))
+                return
+              }
+              // Concatenate new data to the buffer, trimming any
+              // leading/trailing whitespace (the response is a JSON array of
+              // base64-encoded data, meaning that whitespace is not significant)
+              buffer = buffer + Buffer.from(value).toString().trim()
+              // If there was only whitespace, try reading again
+              if (!buffer) break
+              if (state === 'read-new-response') {
+                // Response is in JSON format, so we look for the start of an
+                // array (`[`)
+                if (buffer[0] !== '[') {
+                  controller.error(new Error('Invalid response: no array start delimiter'))
+                  return
+                }
+                // Trim the array start delimiter from the buffer
+                buffer = buffer.slice(1)
+              } else if (state === 'read-eos') {
+                // If in 'read-eos' state and still reading data, it's an error
+                // because the response isn't valid JSON (there should be
+                // nothing other than whitespace after `]`)
+                controller.error(new Error('Invalid data at the end of response'))
+                return
+              }
+              // If not handling new response or end-of-stream, switch to
+              // processing events
+              state = 'events'
+              break
+            }
+            case 'events': {
+              // Process events by looking for a comma or closing bracket that
+              // indicates the end of an event
+              const nextIdx = buffer.search(/(?<=\s*)[,\]]/)
+              // If the end of the event isn't found, go back to reading more
+              // data
+              if (nextIdx < 0) {
+                state = 'read'
+                break
+              }
+              let enqueued = false
+              try {
+                // Extract the current event's value and trim whitespace
+                const eventValue = buffer.slice(0, nextIdx).trim()
+                if (eventValue) {
+                  // Check if the event limit is reached; if so, throw an error
+                  if (count === requestLimit) {
+                    controller.error(new Error('Received too many events'))
```

Comment:
Should this be a special error that's treated differently?

Reply:
Special how?

Reply:
I'm not sure what happens when an error is thrown, hopefully the whole state we've already built up by processing the events doesn't get discarded. If it does get discarded, then detecting a special error could be a way to prevent it from being discarded. (I'm guessing the state isn't thrown out though, so my comment is probably unnecessary)